coders/qoi.c

```diff
@@ -0,0 +1,450 @@
+
+#define QOI_MASK_2    0xc0 /* 11000000 */
+
+#define QOI_COLOR_HASH(C) (C.rgba.r*3 + C.rgba.g*5 + C.rgba.b*7 + C.rgba.a*11)
+
+static unsigned int IsQOI(const unsigned char *magick,const size_t length)
+{
+  if (length < 4)
+    return(MagickFalse);
+  if (LocaleNCompare((char *) magick, "qoif", 4) != 0)
+    return(MagickTrue);
+  return(MagickFalse);
+}
+
+
+typedef union {
+  struct { unsigned char r, g, b, a; } rgba;
+  unsigned int v;
+} qoi_rgba_t;
+
+static Image *ReadQOIImage(const ImageInfo *image_info,
+  ExceptionInfo *exception)
+{
+  Image
+    *image;
+
+  QuantumInfo
+    *quantum_info;
+
+  Quantum
+    *q;
+
+  MagickBooleanType
+    status;
+
+  size_t
+     colorspace,
+     channels;
+
+  ssize_t
+    count,
+    p;
+
+  unsigned char
+    magick[4];
+
+  qoi_rgba_t
+    px,
+    lut[64];
+
+  int
+    b,
+    vg,
+    run;
+
+  /*
+    Open image file.
+  */
+  assert(image_info != (const ImageInfo *) NULL);
+  assert(image_info->signature == MagickCoreSignature);
+  if (image_info->debug != MagickFalse)
+    (void) LogMagickEvent(TraceEvent,GetMagickModule(),"%s",
+      image_info->filename);
+  assert(exception != (ExceptionInfo *) NULL);
+  assert(exception->signature == MagickCoreSignature);
+  image=AcquireImage(image_info,exception);
+  status=OpenBlob(image_info,image,ReadBinaryBlobMode,exception);
+  if (status == MagickFalse)
+    {
+      image=DestroyImageList(image);
+      return((Image *) NULL);
+    }
+  /*
+    Initialize image structure.
+  */
+  image->endian=MSBEndian;
+  image->depth=8;
+  count=ReadBlob(image,4,magick);
+  if ((count != 4) || (LocaleNCompare((char *) magick, "qoif", 4) != 0))
+    ThrowReaderException(CorruptImageError,"ImproperImageHeader");
+  image->columns=(size_t) ReadBlobMSBLong(image);
+  image->rows=(size_t) ReadBlobMSBLong(image);
+  image->alpha_trait=BlendPixelTrait;
+  if (image->columns == 0 || image->rows == 0)
+    ThrowReaderException(CorruptImageError,"NegativeOrZeroImageSize");
+
+  channels=ReadBlobByte(image);
+  if (channels == 3)
+    SetQuantumImageType(image,RGBQuantum);
+  else if (channels == 4)
+    SetQuantumImageType(image,RGBAQuantum);
+  else
+    ThrowReaderException(CorruptImageError,"ImproperImageHeader");
+
+  colorspace=ReadBlobByte(image);
+  if (colorspace == QOI_SRGB)
+    SetImageColorspace(image,sRGBColorspace,exception);
+  else if (colorspace == QOI_LINEAR)
+    SetImageColorspace(image,RGBColorspace,exception);
+  else
+    ThrowReaderException(CorruptImageError,"ImproperImageHeader");
+
+  status=SetImageExtent(image,image->columns,image->rows,exception);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+  quantum_info=AcquireQuantumInfo(image_info,image);
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+  status=SetQuantumFormat(image,quantum_info,UnsignedQuantumFormat);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
```

Comment:
Leaking `quantum_info` here and you don't appear to be using it so we better remove this variable?